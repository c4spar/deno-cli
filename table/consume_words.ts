import { strLength } from "./_utils.ts";

export function consumeWords(length: number, content: string): string {
  let consumed = "";
  const words: Array<string> = content.split("\n")[0]?.split(/ /g);

  for (let i = 0; i < words.length; i++) {
    const word: string = words[i];

    // consume minimum one word
    if (consumed) {
      const nextLength = strLength(word);
      const consumedLength = strLength(consumed);
      if (consumedLength + nextLength >= length) {
        break;
      }
    }

    consumed += (i > 0 ? " " : "") + word;
  }

  return consumed;
}
