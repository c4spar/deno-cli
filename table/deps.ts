export { stripColor } from "https://deno.land/std@0.216.0/fmt/colors.ts";
export { unicodeWidth } from "https://deno.land/std@0.216.0/console/unicode_width.ts";
