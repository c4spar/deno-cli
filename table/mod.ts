export * from "./border.ts";
export * from "./cell.ts";
export * from "./row.ts";
export * from "./table.ts";
