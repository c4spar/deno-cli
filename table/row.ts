import {
  Cell,
  CellType,
  CellValue,
  Direction,
  GetCellValue,
  ValueParser,
} from "./cell.ts";

/** Row type. */
export type RowType<TCell extends CellType> =
  | Row<TCell>
  | ReadonlyArray<TCell>
  | TCell;

export type GetRowInnerValue<TRow extends RowType<CellType>> = TRow extends
  infer TRow
  ? TRow extends ReadonlyArray<infer Value extends CellType>
    ? GetCellValue<Value>
  : TRow
  : never;

export type GetRowValue<TRow extends RowType<CellType>> = TRow extends
  infer TRow ? TRow extends ReadonlyArray<infer Value extends CellType> ? Value
  : TRow
  : never;

export type UnwrapRow<TRow extends RowType<CellType>> = TRow extends infer TRow
  ? TRow extends Row<infer Value extends CellType> ? Value
  : TRow
  : never;

/** @deprecated Use `Record<string, TValue>` instead. */
export type DataRow<TValue extends CellValue> = Record<string, TValue>;

/** Row options. */
export interface RowOptions<
  TValue extends CellValue,
> {
  indent?: number;
  border?: boolean;
  align?: Direction;
  value?: ValueParser<TValue>;
}

/**
 * Row representation.
 *
 * Can be used to customize a single row.
 *
 * ```ts
 * import { Row, Table } from "./mod.ts";
 *
 * new Table()
 *   .body([
 *     new Row("Foo", "Bar").align("right"),
 *     ["Beep", "Boop"],
 *   ])
 *   .render();
 * ```
 */
export class Row<
  TCell extends CellType,
> extends Array<TCell> {
  protected options: RowOptions<GetCellValue<TCell>> = {};

  /**
   * Create a new row. If cells is a row, all cells and options of the row will
   * be copied to the new row.
   *
   * @param value Cells or row.
   */
  public static from<
    TCell extends CellType,
  >(
    value: RowType<TCell>,
  ): Row<TCell> {
    if (Array.isArray(value)) {
      const row = new this<TCell>(...value);
      if (value instanceof Row) {
        row.options = { ...(value as Row<TCell>).options };
      }
      return row;
    }

    return new this(value as TCell);
  }

  /** Clone row recursively with all options. */
  public clone(): this {
    const cells = this.map((cell) =>
      cell instanceof Cell ? cell.clone() : cell
    );
    const row = Row.from(cells) as this;
    row.options = { ...this.options };
    return row;
  }

  /**
   * Setter:
   */

  /**
   * Enable/disable cell border.
   *
   * @param enable    Enable/disable cell border.
   * @param override  Override existing value.
   */
  public border(enable = true, override = true): this {
    if (override || typeof this.options.border === "undefined") {
      this.options.border = enable;
    }
    return this;
  }

  /**
   * Align row content.
   *
   * @param direction Align direction.
   * @param override  Override existing value.
   */
  public align(direction: Direction, override = true): this {
    if (override || typeof this.options.align === "undefined") {
      this.options.align = direction;
    }
    return this;
  }

  /**
   * Register cell value parser.
   * @param fn  Value parser callback function.
   */
  public value(fn: ValueParser<GetCellValue<TCell>>): this {
    this.options.value = fn;
    return this;
  }

  /**
   * Getter:
   */

  /** Check if row has border. */
  public getBorder(): boolean | undefined {
    return this.options.border;
  }

  /** Check if row or any child cell has border. */
  public hasBorder(): boolean {
    return this.getBorder() ||
      this.some((cell) => cell instanceof Cell && cell.getBorder());
  }

  /** Get row alignment. */
  public getAlign(): Direction | undefined {
    return this.options.align;
  }

  /** Get value parser. */
  public getCellValueParser(): ValueParser<GetCellValue<TCell>> | undefined {
    return this.options.value;
  }
}

/** @deprecated Use `RowType` instead. */
export type IRow<TValue extends CellValue> = RowType<TValue>;

/** @deprecated Use `Record<string, TValue>` instead. */
export type IDataRow<TValue extends CellValue> = DataRow<TValue>;
