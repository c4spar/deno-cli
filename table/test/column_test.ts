import { Table } from "../table.ts";
import { assertSnapshot, assertType, IsExact } from "../../dev_deps.ts";
import { Row } from "../row.ts";
import { Cell } from "../cell.ts";

const createTable = () =>
  new Table()
    .header(["Foo", "Bar", "Baz"])
    .body([
      ["foo bar baz", "baz", "beep boop"],
      ["baz", "beep boop", "foo bar baz"],
      ["beep boop", "foo bar baz", "baz"],
    ]);

Deno.test({
  name: "[table] should set border on columns",
  async fn(t) {
    await assertSnapshot(
      t,
      createTable()
        .columns([{
          border: true,
        }, {
          border: false,
        }, {
          border: true,
        }])
        .toString(),
      { serializer },
    );
  },
});

Deno.test({
  name: "[table] should set align on columns",
  async fn(t) {
    await assertSnapshot(
      t,
      createTable()
        .columns([{
          align: "left",
        }, {
          align: "center",
        }, {
          align: "right",
        }])
        .border()
        .toString(),
      { serializer },
    );
  },
});

Deno.test({
  name: "[table] should set width on columns",
  async fn(t) {
    await assertSnapshot(
      t,
      createTable()
        .columns([{
          maxWidth: 4,
        }, {
          minWidth: 20,
        }, {
          align: "right",
        }])
        .border()
        .toString(),
      { serializer },
    );
  },
});

Deno.test({
  name: "[table] should set padding on columns",
  async fn(t) {
    await assertSnapshot(
      t,
      createTable()
        .columns([{
          padding: 5,
          align: "left",
        }, {
          padding: 4,
          align: "center",
        }, {
          padding: 3,
          align: "right",
        }])
        .border()
        .toString(),
      { serializer },
    );
  },
});

Deno.test({
  name: "[table] should set column options with column method",
  async fn(t) {
    const table = createTable();
    table.getColumn(0)?.padding(5);
    table.getColumn(0)?.align("left");

    table.getColumn(1)?.padding(5);
    table.getColumn(1)?.minWidth(20);
    table.getColumn(1)?.align("center");

    table.getColumn(2)?.padding(5);
    table.getColumn(2)?.align("right");

    await assertSnapshot(
      t,
      table
        .border()
        .toString(),
      { serializer },
    );
  },
});

/** Generic type tests */

Deno.test({
  name: "[table] should call headerValue",
  fn: async (t) => {
    await assertSnapshot(
      t,
      new Table()
        .border()
        .header([1, "2", new Date(1000), new RegExp(""), {
          foo: "bar",
        }])
        .columns([{
          headerValue: (value) => {
            assertType<IsExact<typeof value, 1>>(true);
            return `Header 1: ${Deno.inspect(value)}`;
          },
        }, {
          headerValue: (value) => {
            assertType<IsExact<typeof value, "2">>(true);
            return `Header 2: ${Deno.inspect(value)}`;
          },
        }, {
          headerValue: (value) => {
            assertType<IsExact<typeof value, Date>>(true);
            return `Header 3: ${Deno.inspect(value)}`;
          },
        }, {
          headerValue: (value) => {
            assertType<IsExact<typeof value, RegExp>>(true);
            return `Header 4: ${Deno.inspect(value)}`;
          },
        }, {
          headerValue: (value) => {
            assertType<IsExact<typeof value, { readonly foo: "bar" }>>(true);
            return `Header 5: ${Deno.inspect(value)}`;
          },
        }])
        .toString(),
      { serializer },
    );
  },
});

Deno.test({
  name: "[table] should call headerValue and value",
  fn: async (t) => {
    await assertSnapshot(
      t,
      new Table()
        .border()
        .header([1, "2", new Date(1000), new RegExp(""), { foo: "bar" }])
        .body([
          ["1", 2, 3, { beep: true }, [1]],
          ["1", 2, "3", { beep: true }, [1]],
          ["1", 2, 3, { beep: true }, [1]],
        ])
        .columns([{
          headerValue: (value) => {
            assertType<IsExact<typeof value, 1>>(true);
            return `Header 1: ${Deno.inspect(value)}`;
          },
          value: (value) => {
            assertType<IsExact<typeof value, "1">>(true);
            return `Body 1: ${Deno.inspect(value)}`;
          },
        }, {
          headerValue: (value) => {
            assertType<IsExact<typeof value, "2">>(true);
            return `Header 2: ${Deno.inspect(value)}`;
          },
          value: (value) => {
            assertType<IsExact<typeof value, 2>>(true);
            return `Body 2: ${Deno.inspect(value)}`;
          },
        }, {
          headerValue: (value) => {
            assertType<IsExact<typeof value, Date>>(true);
            return `Header 3: ${Deno.inspect(value)}`;
          },
          value: (value) => {
            assertType<IsExact<typeof value, 3 | "3">>(true);
            return `Body 3: ${Deno.inspect(value)}`;
          },
        }, {
          headerValue: (value) => {
            assertType<IsExact<typeof value, RegExp>>(true);
            return `Header 4: ${Deno.inspect(value)}`;
          },
          value: (value) => {
            assertType<IsExact<typeof value, { readonly beep: true }>>(true);
            return `Body 4: ${Deno.inspect(value)}`;
          },
        }, {
          headerValue: (value) => {
            assertType<IsExact<typeof value, { readonly foo: "bar" }>>(true);
            return `Header 5: ${Deno.inspect(value)}`;
          },
          value: (value) => {
            assertType<IsExact<typeof value, readonly [1]>>(true);
            return `Body 5: ${Deno.inspect(value)}`;
          },
        }])
        .toString(),
      { serializer },
    );
  },
});

Deno.test({
  name: "[table] should call value",
  fn: async (t) => {
    await assertSnapshot(
      t,
      new Table()
        .border()
        .body([
          ["1", 2, 3, { beep: true }, [1]],
          ["1", 2, "3", { beep: true }, [1]],
          ["1", 2, 3, { beep: true }, [1]],
        ])
        .columns([{
          headerValue: (value) => {
            assertType<IsExact<typeof value, unknown>>(true);
            return `Header 1: ${Deno.inspect(value)}`;
          },
          value: (value) => {
            assertType<IsExact<typeof value, "1">>(true);
            return `Header 1: ${Deno.inspect(value)}`;
          },
        }, {
          headerValue: (value) => {
            assertType<IsExact<typeof value, unknown>>(true);
            return `Header 2: ${Deno.inspect(value)}`;
          },
          value: (value) => {
            assertType<IsExact<typeof value, 2>>(true);
            return `Body 2: ${Deno.inspect(value)}`;
          },
        }, {
          headerValue: (value) => {
            assertType<IsExact<typeof value, unknown>>(true);
            return `Header 3: ${Deno.inspect(value)}`;
          },
          value: (value) => {
            assertType<IsExact<typeof value, 3 | "3">>(true);
            return `Body 3: ${Deno.inspect(value)}`;
          },
        }, {
          headerValue: (value) => {
            assertType<IsExact<typeof value, unknown>>(true);
            return `Header 4: ${Deno.inspect(value)}`;
          },
          value: (value) => {
            assertType<IsExact<typeof value, { readonly beep: true }>>(true);
            return `Body 4: ${Deno.inspect(value)}`;
          },
        }, {
          headerValue: (value) => {
            assertType<IsExact<typeof value, unknown>>(true);
            return `Header 5: ${Deno.inspect(value)}`;
          },
          value: (value) => {
            assertType<IsExact<typeof value, readonly [1]>>(true);
            return `Body 5: ${Deno.inspect(value)}`;
          },
        }])
        .toString(),
      { serializer },
    );
  },
});

Deno.test({
  name: "[table] should call value and headerValue with json data",
  fn: async (t) => {
    type Data =
      | {
        readonly firstName: "Gino";
        readonly lastName: "Aicheson";
        readonly age: 21;
        readonly email: "gaicheson0@nydailynews.com";
      }
      | {
        readonly firstName: "Godfry";
        readonly lastName: "Pedycan";
        readonly age: 33;
        readonly email: "gpedycan1@state.gov";
      }
      | {
        readonly firstName: "Loni";
        readonly lastName: "Miller";
        readonly age: 24;
        readonly email: "lmiller2@chron.com";
      };

    await assertSnapshot(
      t,
      new Table()
        .border()
        .body([
          {
            firstName: "Gino",
            lastName: "Aicheson",
            age: 21,
            email: "gaicheson0@nydailynews.com",
          },
          {
            firstName: "Godfry",
            lastName: "Pedycan",
            age: 33,
            email: "gpedycan1@state.gov",
          },
          {
            firstName: "Loni",
            lastName: "Miller",
            age: 24,
            email: "lmiller2@chron.com",
          },
        ])
        .columns([{
          headerValue: (value) => {
            assertType<IsExact<typeof value, unknown>>(true);
            return `Header 1: ${Deno.inspect(value)}`;
          },
          value: (value) => {
            assertType<IsExact<typeof value, Data>>(true);
            return `Body 1: ${Deno.inspect(value)}`;
          },
        }, {
          headerValue: (value) => {
            assertType<IsExact<typeof value, unknown>>(true);
            return `Header 2: ${Deno.inspect(value)}`;
          },
          value: (value) => {
            assertType<IsExact<typeof value, Data>>(true);
            return `Body 2: ${Deno.inspect(value)}`;
          },
        }, {
          headerValue: (value) => {
            assertType<IsExact<typeof value, unknown>>(true);
            return `Header 3: ${Deno.inspect(value)}`;
          },
          value: (value) => {
            assertType<IsExact<typeof value, Data>>(true);
            return `Body 3: ${Deno.inspect(value)}`;
          },
        }, {
          headerValue: (value) => {
            assertType<IsExact<typeof value, unknown>>(true);
            return `Header 4: ${Deno.inspect(value)}`;
          },
          value: (value) => {
            assertType<IsExact<typeof value, Data>>(true);
            return `Body 4: ${Deno.inspect(value)}`;
          },
        }])
        .toString(),
      { serializer },
    );
  },
});

Deno.test({
  name: "[table] should call value and headerValue with json data and header",
  fn: async (t) => {
    type Data =
      | {
        readonly firstName: "Gino";
        readonly lastName: "Aicheson";
        readonly age: 21;
        readonly email: "gaicheson0@nydailynews.com";
      }
      | {
        readonly firstName: "Godfry";
        readonly lastName: "Pedycan";
        readonly age: 33;
        readonly email: "gpedycan1@state.gov";
      }
      | {
        readonly firstName: "Loni";
        readonly lastName: "Miller";
        readonly age: 24;
        readonly email: "lmiller2@chron.com";
      };

    await assertSnapshot(
      t,
      new Table()
        .border()
        .header([1, "2", 3, "4"])
        .body([
          {
            firstName: "Gino",
            lastName: "Aicheson",
            age: 21,
            email: "gaicheson0@nydailynews.com",
          },
          {
            firstName: "Godfry",
            lastName: "Pedycan",
            age: 33,
            email: "gpedycan1@state.gov",
          },
          {
            firstName: "Loni",
            lastName: "Miller",
            age: 24,
            email: "lmiller2@chron.com",
          },
        ])
        .columns([{
          headerValue: (value) => {
            assertType<IsExact<typeof value, 1>>(true);
            return `Header 1: ${Deno.inspect(value)}`;
          },
          value: (value) => {
            assertType<IsExact<typeof value, Data>>(true);
            return `Body 1: ${Deno.inspect(value)}`;
          },
        }, {
          headerValue: (value) => {
            assertType<IsExact<typeof value, "2">>(true);
            return `Header 2: ${Deno.inspect(value)}`;
          },
          value: (value) => {
            assertType<IsExact<typeof value, Data>>(true);
            return `Body 2: ${Deno.inspect(value)}`;
          },
        }, {
          headerValue: (value) => {
            assertType<IsExact<typeof value, 3>>(true);
            return `Header 3: ${Deno.inspect(value)}`;
          },
          value: (value) => {
            assertType<IsExact<typeof value, Data>>(true);
            return `Body 3: ${Deno.inspect(value)}`;
          },
        }, {
          headerValue: (value) => {
            assertType<IsExact<typeof value, "4">>(true);
            return `Header 4: ${Deno.inspect(value)}`;
          },
          value: (value) => {
            assertType<IsExact<typeof value, Data>>(true);
            return `Body 4: ${Deno.inspect(value)}`;
          },
        }])
        .toString(),
      { serializer },
    );
  },
});

Deno.test({
  name: "[table] should call headerValue and value with Row class",
  fn: async (t) => {
    type ExpectedType = number | string | Date | RegExp | { foo: string };

    await assertSnapshot(
      t,
      new Table()
        .border()
        .header(
          Row.from(
            [1, "2", new Date(1000), new Cell(new RegExp("")), {
              foo: "bar",
            }],
          ),
        )
        .body([
          Row.from(
            [1, "2", new Date(1000), new Cell(new RegExp("")), {
              foo: "bar",
            }],
          ),
        ])
        .columns([{
          headerValue: (value) => {
            assertType<IsExact<typeof value, ExpectedType>>(true);
            return `Header 1: ${Deno.inspect(value)}`;
          },
          value: (value) => {
            assertType<IsExact<typeof value, ExpectedType>>(true);
            return `Body 1: ${Deno.inspect(value)}`;
          },
        }, {
          headerValue: (value) => {
            assertType<IsExact<typeof value, ExpectedType>>(true);
            return `Header 2: ${Deno.inspect(value)}`;
          },
          value: (value) => {
            assertType<IsExact<typeof value, ExpectedType>>(true);
            return `Body 2: ${Deno.inspect(value)}`;
          },
        }, {
          headerValue: (value) => {
            assertType<IsExact<typeof value, ExpectedType>>(true);
            return `Header 3: ${Deno.inspect(value)}`;
          },
          value: (value) => {
            assertType<IsExact<typeof value, ExpectedType>>(true);
            return `Body 3: ${Deno.inspect(value)}`;
          },
        }, {
          headerValue: (value) => {
            assertType<IsExact<typeof value, ExpectedType>>(true);
            return `Header 4: ${Deno.inspect(value)}`;
          },
          value: (value) => {
            assertType<IsExact<typeof value, ExpectedType>>(true);
            return `Body 4: ${Deno.inspect(value)}`;
          },
        }, {
          headerValue: (value) => {
            assertType<IsExact<typeof value, ExpectedType>>(true);
            return `Header 5: ${Deno.inspect(value)}`;
          },
          value: (value) => {
            assertType<IsExact<typeof value, ExpectedType>>(true);
            return `Body 5: ${Deno.inspect(value)}`;
          },
        }])
        .toString(),
      { serializer },
    );
  },
});

Deno.test({
  name:
    "[table] should allow header with array and body with array and row class",
  fn: async (t) => {
    type ExpectedType =
      | number
      | string
      | Date
      | RegExp
      | Map<string, number>
      | { foo: string };

    const table = new Table()
      .border()
      .header(
        [1, "2", new Date(1000), new Cell(new RegExp("")), {
          foo: "bar",
        }],
      )
      .body([
        Row.from(
          [1, "2", new Date(1000), new Cell(new RegExp("")), {
            foo: "bar",
          }],
        ),
        [
          new Map<string, number>(),
          "2",
          new Date(1000),
          new Cell(new RegExp("")),
          {
            foo: "bar",
          },
        ],
      ])
      .columns([{
        headerValue: (value) => {
          assertType<IsExact<typeof value, 1>>(true);
          return `Header 1: ${Deno.inspect(value)}`;
        },
        value: (value) => {
          assertType<IsExact<typeof value, ExpectedType>>(true);
          return `Body 1: ${Deno.inspect(value)}`;
        },
      }, {
        headerValue: (value) => {
          assertType<IsExact<typeof value, "2">>(true);
          return `Header 2: ${Deno.inspect(value)}`;
        },
        value: (value) => {
          assertType<IsExact<typeof value, ExpectedType>>(true);
          return `Body 2: ${Deno.inspect(value)}`;
        },
      }, {
        headerValue: (value) => {
          assertType<IsExact<typeof value, Date>>(true);
          return `Header 3: ${Deno.inspect(value)}`;
        },
        value: (value) => {
          assertType<IsExact<typeof value, ExpectedType>>(true);
          return `Body 3: ${Deno.inspect(value)}`;
        },
      }, {
        headerValue: (value) => {
          assertType<IsExact<typeof value, RegExp>>(true);
          return `Header 4: ${Deno.inspect(value)}`;
        },
        value: (value) => {
          assertType<IsExact<typeof value, ExpectedType>>(true);
          return `Body 4: ${Deno.inspect(value)}`;
        },
      }, {
        headerValue: (value) => {
          assertType<IsExact<typeof value, { readonly foo: "bar" }>>(true);
          return `Header 5: ${Deno.inspect(value)}`;
        },
        value: (value) => {
          assertType<IsExact<typeof value, ExpectedType>>(true);
          return `Body 5: ${Deno.inspect(value)}`;
        },
      }]);

    await assertSnapshot(
      t,
      table.toString(),
      { serializer },
    );
  },
});

function serializer<T>(value: T): T {
  return value;
}
