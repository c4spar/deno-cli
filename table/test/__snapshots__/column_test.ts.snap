export const snapshot = {};

snapshot[`[table] should set border on columns 1`] = `
  Foo           Bar           Baz          
┌─────────────┐             ┌─────────────┐
│ foo bar baz │ baz         │ beep boop   │
├─────────────┤             ├─────────────┤
│ baz         │ beep boop   │ foo bar baz │
├─────────────┤             ├─────────────┤
│ beep boop   │ foo bar baz │ baz         │
└─────────────┘             └─────────────┘
`;

snapshot[`[table] should set align on columns 1`] = `
┌─────────────┬─────────────┬─────────────┐
│ Foo         │ Bar         │ Baz         │
├─────────────┼─────────────┼─────────────┤
│ foo bar baz │     baz     │   beep boop │
├─────────────┼─────────────┼─────────────┤
│ baz         │  beep boop  │ foo bar baz │
├─────────────┼─────────────┼─────────────┤
│ beep boop   │ foo bar baz │         baz │
└─────────────┴─────────────┴─────────────┘
`;

snapshot[`[table] should set width on columns 1`] = `
┌──────┬──────────────────────┬─────────────┐
│ Foo  │ Bar                  │ Baz         │
├──────┼──────────────────────┼─────────────┤
│ foo  │ baz                  │   beep boop │
│ bar  │                      │             │
│ baz  │                      │             │
├──────┼──────────────────────┼─────────────┤
│ baz  │ beep boop            │ foo bar baz │
├──────┼──────────────────────┼─────────────┤
│ beep │ foo bar baz          │         baz │
│ boop │                      │             │
└──────┴──────────────────────┴─────────────┘
`;

snapshot[`[table] should set padding on columns 1`] = `
┌─────────────────────┬───────────────────┬─────────────────┐
│     Foo             │    Bar            │   Baz           │
├─────────────────────┼───────────────────┼─────────────────┤
│     foo bar baz     │        baz        │     beep boop   │
├─────────────────────┼───────────────────┼─────────────────┤
│     baz             │     beep boop     │   foo bar baz   │
├─────────────────────┼───────────────────┼─────────────────┤
│     beep boop       │    foo bar baz    │           baz   │
└─────────────────────┴───────────────────┴─────────────────┘
`;

snapshot[`[table] should set column options with column method 1`] = `
┌─────────────────────┬──────────────────────────────┬─────────────────────┐
│     Foo             │     Bar                      │     Baz             │
├─────────────────────┼──────────────────────────────┼─────────────────────┤
│     foo bar baz     │             baz              │       beep boop     │
├─────────────────────┼──────────────────────────────┼─────────────────────┤
│     baz             │          beep boop           │     foo bar baz     │
├─────────────────────┼──────────────────────────────┼─────────────────────┤
│     beep boop       │         foo bar baz          │             baz     │
└─────────────────────┴──────────────────────────────┴─────────────────────┘
`;

snapshot[`[table] should call headerValue 1`] = `
┌─────────────┬───────────────┬────────────────────────────────────┬──────────────────┬──────────────────────────┐
│ Header 1: 1 │ Header 2: "2" │ Header 3: 1970-01-01T00:00:01.000Z │ Header 4: /(?:)/ │ Header 5: { foo: "bar" } │
└─────────────┴───────────────┴────────────────────────────────────┴──────────────────┴──────────────────────────┘
`;

snapshot[`[table] should call headerValue and value 1`] = `
┌─────────────┬───────────────┬────────────────────────────────────┬────────────────────────┬──────────────────────────┐
│ Header 1: 1 │ Header 2: "2" │ Header 3: 1970-01-01T00:00:01.000Z │ Header 4: /(?:)/       │ Header 5: { foo: "bar" } │
├─────────────┼───────────────┼────────────────────────────────────┼────────────────────────┼──────────────────────────┤
│ Body 1: "1" │ Body 2: 2     │ Body 3: 3                          │ Body 4: { beep: true } │ Body 5: [ 1 ]            │
├─────────────┼───────────────┼────────────────────────────────────┼────────────────────────┼──────────────────────────┤
│ Body 1: "1" │ Body 2: 2     │ Body 3: "3"                        │ Body 4: { beep: true } │ Body 5: [ 1 ]            │
├─────────────┼───────────────┼────────────────────────────────────┼────────────────────────┼──────────────────────────┤
│ Body 1: "1" │ Body 2: 2     │ Body 3: 3                          │ Body 4: { beep: true } │ Body 5: [ 1 ]            │
└─────────────┴───────────────┴────────────────────────────────────┴────────────────────────┴──────────────────────────┘
`;

snapshot[`[table] should call value 1`] = `
┌───────────────┬───────────┬─────────────┬────────────────────────┬───────────────┐
│ Header 1: "1" │ Body 2: 2 │ Body 3: 3   │ Body 4: { beep: true } │ Body 5: [ 1 ] │
├───────────────┼───────────┼─────────────┼────────────────────────┼───────────────┤
│ Header 1: "1" │ Body 2: 2 │ Body 3: "3" │ Body 4: { beep: true } │ Body 5: [ 1 ] │
├───────────────┼───────────┼─────────────┼────────────────────────┼───────────────┤
│ Header 1: "1" │ Body 2: 2 │ Body 3: 3   │ Body 4: { beep: true } │ Body 5: [ 1 ] │
└───────────────┴───────────┴─────────────┴────────────────────────┴───────────────┘
`;

snapshot[`[table] should call value and headerValue with json data 1`] = `
┌───────────────────────────────────────┬───────────────────────────────────────┬───────────────────────────────────────┬───────────────────────────────────────┐
│ Body 1: {                             │ Body 2: {                             │ Body 3: {                             │ Body 4: {                             │
│   firstName: "Gino",                  │   firstName: "Gino",                  │   firstName: "Gino",                  │   firstName: "Gino",                  │
│   lastName: "Aicheson",               │   lastName: "Aicheson",               │   lastName: "Aicheson",               │   lastName: "Aicheson",               │
│   age: 21,                            │   age: 21,                            │   age: 21,                            │   age: 21,                            │
│   email: "gaicheson0@nydailynews.com" │   email: "gaicheson0@nydailynews.com" │   email: "gaicheson0@nydailynews.com" │   email: "gaicheson0@nydailynews.com" │
│ }                                     │ }                                     │ }                                     │ }                                     │
├───────────────────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┤
│ Body 1: {                             │ Body 2: {                             │ Body 3: {                             │ Body 4: {                             │
│   firstName: "Godfry",                │   firstName: "Godfry",                │   firstName: "Godfry",                │   firstName: "Godfry",                │
│   lastName: "Pedycan",                │   lastName: "Pedycan",                │   lastName: "Pedycan",                │   lastName: "Pedycan",                │
│   age: 33,                            │   age: 33,                            │   age: 33,                            │   age: 33,                            │
│   email: "gpedycan1@state.gov"        │   email: "gpedycan1@state.gov"        │   email: "gpedycan1@state.gov"        │   email: "gpedycan1@state.gov"        │
│ }                                     │ }                                     │ }                                     │ }                                     │
├───────────────────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┤
│ Body 1: {                             │ Body 2: {                             │ Body 3: {                             │ Body 4: {                             │
│   firstName: "Loni",                  │   firstName: "Loni",                  │   firstName: "Loni",                  │   firstName: "Loni",                  │
│   lastName: "Miller",                 │   lastName: "Miller",                 │   lastName: "Miller",                 │   lastName: "Miller",                 │
│   age: 24,                            │   age: 24,                            │   age: 24,                            │   age: 24,                            │
│   email: "lmiller2@chron.com"         │   email: "lmiller2@chron.com"         │   email: "lmiller2@chron.com"         │   email: "lmiller2@chron.com"         │
│ }                                     │ }                                     │ }                                     │ }                                     │
└───────────────────────────────────────┴───────────────────────────────────────┴───────────────────────────────────────┴───────────────────────────────────────┘
`;

snapshot[`[table] should call value and headerValue with json data and header 1`] = `
┌───────────────────────────────────────┬───────────────────────────────────────┬───────────────────────────────────────┬───────────────────────────────────────┐
│ Header 1: 1                           │ Header 2: "2"                         │ Header 3: 3                           │ Header 4: "4"                         │
├───────────────────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┤
│ Body 1: {                             │ Body 2: {                             │ Body 3: {                             │ Body 4: {                             │
│   firstName: "Gino",                  │   firstName: "Gino",                  │   firstName: "Gino",                  │   firstName: "Gino",                  │
│   lastName: "Aicheson",               │   lastName: "Aicheson",               │   lastName: "Aicheson",               │   lastName: "Aicheson",               │
│   age: 21,                            │   age: 21,                            │   age: 21,                            │   age: 21,                            │
│   email: "gaicheson0@nydailynews.com" │   email: "gaicheson0@nydailynews.com" │   email: "gaicheson0@nydailynews.com" │   email: "gaicheson0@nydailynews.com" │
│ }                                     │ }                                     │ }                                     │ }                                     │
├───────────────────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┤
│ Body 1: {                             │ Body 2: {                             │ Body 3: {                             │ Body 4: {                             │
│   firstName: "Godfry",                │   firstName: "Godfry",                │   firstName: "Godfry",                │   firstName: "Godfry",                │
│   lastName: "Pedycan",                │   lastName: "Pedycan",                │   lastName: "Pedycan",                │   lastName: "Pedycan",                │
│   age: 33,                            │   age: 33,                            │   age: 33,                            │   age: 33,                            │
│   email: "gpedycan1@state.gov"        │   email: "gpedycan1@state.gov"        │   email: "gpedycan1@state.gov"        │   email: "gpedycan1@state.gov"        │
│ }                                     │ }                                     │ }                                     │ }                                     │
├───────────────────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┤
│ Body 1: {                             │ Body 2: {                             │ Body 3: {                             │ Body 4: {                             │
│   firstName: "Loni",                  │   firstName: "Loni",                  │   firstName: "Loni",                  │   firstName: "Loni",                  │
│   lastName: "Miller",                 │   lastName: "Miller",                 │   lastName: "Miller",                 │   lastName: "Miller",                 │
│   age: 24,                            │   age: 24,                            │   age: 24,                            │   age: 24,                            │
│   email: "lmiller2@chron.com"         │   email: "lmiller2@chron.com"         │   email: "lmiller2@chron.com"         │   email: "lmiller2@chron.com"         │
│ }                                     │ }                                     │ }                                     │ }                                     │
└───────────────────────────────────────┴───────────────────────────────────────┴───────────────────────────────────────┴───────────────────────────────────────┘
`;

snapshot[`[table] should call headerValue and value with Row class 1`] = `
┌─────────────┬───────────────┬────────────────────────────────────┬──────────────────┬──────────────────────────┐
│ Header 1: 1 │ Header 2: "2" │ Header 3: 1970-01-01T00:00:01.000Z │ Header 4: /(?:)/ │ Header 5: { foo: "bar" } │
├─────────────┼───────────────┼────────────────────────────────────┼──────────────────┼──────────────────────────┤
│ Body 1: 1   │ Body 2: "2"   │ Body 3: 1970-01-01T00:00:01.000Z   │ Body 4: /(?:)/   │ Body 5: { foo: "bar" }   │
└─────────────┴───────────────┴────────────────────────────────────┴──────────────────┴──────────────────────────┘
`;

snapshot[`[table] should allow header with array and body with array and row class 1`] = `
┌───────────────────┬───────────────┬────────────────────────────────────┬──────────────────┬──────────────────────────┐
│ Header 1: 1       │ Header 2: "2" │ Header 3: 1970-01-01T00:00:01.000Z │ Header 4: /(?:)/ │ Header 5: { foo: "bar" } │
├───────────────────┼───────────────┼────────────────────────────────────┼──────────────────┼──────────────────────────┤
│ Body 1: 1         │ Body 2: "2"   │ Body 3: 1970-01-01T00:00:01.000Z   │ Body 4: /(?:)/   │ Body 5: { foo: "bar" }   │
├───────────────────┼───────────────┼────────────────────────────────────┼──────────────────┼──────────────────────────┤
│ Body 1: Map(0) {} │ Body 2: "2"   │ Body 3: 1970-01-01T00:00:01.000Z   │ Body 4: /(?:)/   │ Body 5: { foo: "bar" }   │
└───────────────────┴───────────────┴────────────────────────────────────┴──────────────────┴──────────────────────────┘
`;
